flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -1011,10 +989,25 @@ TimerService getTimerService() {
 		return timerService;
 	}
 
-	public ProcessingTimeService getProcessingTimeService(int operatorIndex) {
+	@VisibleForTesting
+	StreamOperator<?> getHeadOperator() {
+		return operatorChain.getHeadOperator();
+	}
+
+	public ProcessingTimeService getProcessingTimeService(OperatorID operatorID) {
+		Preconditions.checkNotNull(operatorID);
 		Preconditions.checkState(timerService != null, "The timer service has not been initialized.");
-		MailboxExecutor mailboxExecutor = mailboxProcessor.getMailboxExecutor(operatorIndex);
-		return new ProcessingTimeServiceImpl(timerService, callback -> deferCallbackToMailbox(mailboxExecutor, callback));
+		Preconditions.checkState(operatorChain != null, "operatorChain has not been initialized.");
+
+		ProcessingTimeService processingTimeService = operatorChain.getOperatorProcessingTimeService(operatorID);
+		if (processingTimeService == null) {
+			processingTimeService = new ProcessingTimeServiceImpl(
+				timerService,
+				callback -> deferCallbackToMailbox(operatorChain.getOperatorMailboxExecutor(operatorID), callback));
+			operatorChain.setOperatorProcessingTimeService(operatorID, (ProcessingTimeServiceImpl) processingTimeService);
+		}
```

Comment:
The call chain seems quite strange:
```
operator -> task.getProcessingTimerService(this.getOperatorId)
task -> chain.getProcessingTimerService(id)
chain -> this.getWrapper(id).getTimerService
wrapper -> return this.timerService
```

But it's the `operator` and its `wrapper` that actually need this `timerService`, right?

Also, nullable `getOperatorProcessingTimeService` which is set outside is quite fragile.

So let's store `timerService` in `AbstractStreamOperator` and expose it to its `wrapper`.
Then we also don't need mapping in `OperatorChain` from `OperatorID` to wrapper (`allOperatorWrappers`).
The only issue is to create `ProcessingTimeService` - we can pass some factory for this.

Reply:
Yes, it is true strange here.

Hmm, I have thought about using `AbstractStreamOperator ` to store, but the problem is that `StreamTask` and `OperatorChain` can only see `StreamOperator` , but not `AbstractStreamOperator`

Reply:
And neither `StreamTask` nor `OperatorChain` need `ProcessingTimeService` - they only return it back to `StreamOperator`.

Reply:
The `StreamOperator` interface doesn't have the `#getProcessingTimeService` method, only `AbstractStreamOperator` does.  I guess I don't get what you mean :) .

Reply:
We could add something like `@PublicEvolving interface ProcessingTimersAware { set/get }` and make `AbstractStreamOperator` implement it.

Reply:
I understand it's a bit like `YieldingOperatorFactory`, right? Then we should remove the `StreamTask# getProcessingTimeService`, otherwise we can't ensure that only one `ProcessingTimeService` is bound to one operator?

Reply:
Yes, similar to `YieldingOperatorFactory`. Right.